crates/runix/src/flake_ref/mod.rs

```diff
@@ -92,6 +132,150 @@ impl FromStr for FlakeRef {
+    ///
+    /// Nix supports referring to local files/paths without an explicit scheme.
+    /// However, it distinguishes between flakes in git repositories
+    /// (disambiguated using `git+file://`) from "free" ones (`path:`).
+    ///
+    /// The logic for this in nix can be found at <https://github.com/NixOS/nix/blob/bf7dc3c7dc24f75fa623135750e8f10b8bcd94f9/src/libexpr/flake/flakeref.cc#L112C5-L197>
+    ///
+    /// This method implements a subset of this logic assuming we require the path to exist and to be a flake.
+    ///
+    /// In abstract we
+    ///
+    /// 1. locate a flake (by finding a `flake.nix` in all ancestors)
+    ///   - fail if directory does not exist
+    ///   - fail if skipping file system boundaries
+    ///   - fail if "exiting" a git repository (if there is one)
+    /// 2. locate the root of the "containing" git repo (if applicable)
+    ///   - check whether `.git/shallow` exists and track in the url params
+    ///   - ensure there is no `?dir=` param if flake is in a subdir
+    /// 3. construct a `file+git:` or `path:` url as required
+    pub fn resolve_local(s: impl AsRef<str>) -> Result<Url, ResolveLocalRefError> {
+        let s = s.as_ref();
+        let mut git_url =
+            Url::parse(&format!("git+file://{s}")).map_err(ResolveLocalRefError::ParseUrl)?;
+
+        let path = Path::new(git_url.path());
+        let path = path
+            .canonicalize()
+            .map_err(|err| ResolveLocalRefError::Canonicalize(path.to_path_buf(), err))?;
+
+        // update base to canonicalized path
+        git_url.set_path(&path.to_string_lossy());
+
+        let original_device = path
+            .metadata()
+            .map_err(|err| ResolveLocalRefError::Metadata(path.to_path_buf(), err))?
+            .dev();
+
+        let mut ancestors = path.ancestors();
+
+        let flake_root = loop {
+            let ancestor = ancestors
+                .next()
+                .ok_or_else(|| ResolveLocalRefError::NotFound(path.to_path_buf()))?;
+
+            debug!("looking for flake in {ancestor:?}");
+
+            {
+                let device = ancestor
+                    .metadata()
+                    .map_err(|err| ResolveLocalRefError::Metadata(path.to_path_buf(), err))?
+                    .dev();
+
+                if device != original_device {
+                    Err(ResolveLocalRefError::FileSystemBoundary(
+                        ancestor.to_path_buf(),
+                    ))?
+                }
+            }
+
+            if ancestor.join("flake.nix").exists() {
+                debug!("found for flake in {ancestor:?}");
+                break ancestor;
+            }
+
+            if path.join(".git").exists() {
+                Err(ResolveLocalRefError::GitRepoBoundary(
+                    ancestor.to_path_buf(),
+                ))?;
+            }
+        };
+
+        // if there is git
+        if let Some(git_root) = flake_root.ancestors().find(|ancestor| {
```

Comment:
Optional: my intuition is that having a single loop would be more readable than looping through `ancestors()` twice

Reply:
Again, this mirrors the nix implementation. I thought it might actually be easier to read this way because you separate the flake "finding" from the git "finding". It's a bit confusion that we are also looking for the git boundary during the flake search but well..
Btw. the two loops are already "kind of" a single loop as the full set of ancestors is iterated until we find a flake and then possibly completed when we search the git root.

Reply:
Nope I don't mind